src/tree-sitter-language-mode.js

```diff
@@ -348,25 +349,28 @@ class TreeSitterLanguageMode {
   Section - Syntax Tree APIs
   */
 
-  getRangeForSyntaxNodeContainingRange (range) {
+  getRangeForSyntaxNodeContainingRange (range, selector) {
```

Comment:
How do you imagine the compatibility working for this? The legacy TextMate selectors often target multiple scopes like `string.quoted`, but in a Tree-sitter syntax tree, a node never has more than one `type`; strings for example are just `string`.

I'd imagine that we would want the selector `.string` to match a `string` node, but what do we do about `.string.quoted`? It seems like there's no way around changing callers to provide new selectors. Did you envision changing `toggle-quotes` to just pass `.string` instead of `.string.quoted`?

Reply:
Yeah, the backcompat story here is a bummer.

I did actually change `toggle-quotes` to do that, and it... well, it works for that use case with the JS grammar ;) Perhaps even for most grammars—I imagine the C family and Ruby grammars call their quoted strings `string`s. HTML, meanwhile, does not, calling its text `text` IIRC. (OTOH, HTML has `attribute_value`s, and that's a bummer, because it means that we can't toggle quotes on those (though we couldn't before, either)).

One alternative I considered is looking up the scope map and querying on those. But it would be nice if we could start divesting ourselves of the TextMate CSS classes, particularly for this query interface. We *should* provide a more expressive query interface than just matching on node type, though—ideally, what `toggle-quotes` could do is say, "find me the closest node starting with `"|'` and ending with `"|'`", which would be quite doable.

I guess the question is, how bad is it to semi-break the behavior of `bufferRangeForNodeInRange`? I mean, it was completely broken before, and this is the first time someone's noticed.

Reply:
> We should provide a more expressive query interface than just matching on node type, though—ideally, what toggle-quotes could do is say, "find me the closest node starting with "|' and ending with "|'", which would be quite doable.

Yeah, you're right. That's the logic that `toggle-quotes` should be using; it probably shouldn't rely on the names of any syntax nodes.

But I don't necessarily think we need to build any special query language for finding that syntax node; we should just allow the package to directly inspect the syntax nodes and walk up the tree using `SyntaxNode.parent`. What do you think of adding a new experimental API called something like `syntaxNodeForPosition()` or `syntaxNodeForRange()` that `toggle-quotes` can use to inspect the syntax tree manually?

Reply:
Thing is, there are potentially multiple trees at that point, and the deepest one may not be right.  Consider toggling quotes inside the rhs of `document.body.innerHTML = "<span>hello</span>"`.

But this change makes it quite easy for us to accept a node predicate function as a selector, so the language mode can still handle walking the trees and finding the smallest matching node, and extensions can pass in a function describing the node they want.

Updated to support and test that behavior.

Reply:
On second though, it would be nice to be able to query for support of the new behavior, so I did add `getSyntaxNodeForRange` and `getSyntaxNodeForPosition`. (`bufferRangeForScopeAtPosition` also supports match functions)

It works for toggle-quotes.

Reply:
(See https://github.com/atom/toggle-quotes/pull/58)